tracing-subscriber/src/filter/layer_filters.rs

```diff
@@ -226,31 +386,197 @@ impl<S, F> FilterFn<S, F>
 where
     F: Fn(&Metadata<'_>, &Context<'_, S>) -> bool,
 {
+    /// Constructs a [`Filter`] from a function or closure that returns `true`
+    /// if a span or event should be enabled.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tracing_subscriber::{
+    ///     layer::{Layer, SubscriberExt},
+    ///     filter::FilterFn,
+    ///     util::SubscriberInitExt,
+    /// };
+    ///
+    /// let my_filter = FilterFn::new(|metadata, _| {
+    ///     // Only enable spans or events with the target "interesting_target"
+    ///     metadata.target() == "interesting_target"
+    /// });
+    ///
+    /// let my_layer = // ...
+    ///     # tracing_subscriber::filter::LevelFilter::INFO;
+    ///
+    /// tracing_subscriber::registry()
+    ///     .with(my_layer.with_filter(my_filter))
+    ///     .init();
+    /// ```
     pub fn new(enabled: F) -> Self {
         Self {
             enabled,
             register_callsite: None,
             max_level_hint: None,
+            cacheable: false,
             _s: PhantomData,
         }
     }
+
+    /// Indicates that the result of the [`enabled`] can be cached.
+    ///
+    /// By default, a `FilterFn` assumes that the supplied [`Filter::enabled`]
+    /// function _may_ be dynamic: it may decide whether a given span or event
+    /// is enabled based not only on its [`Metadata`], but also based on the
+    /// current [span context] in which a particular instance of that event
+    /// occurs. Therefore, the default [`Filter::callsite_enabled`]
+    /// implementation will always return [`Interest::sometimes`], indicating
+    /// that the [`Filter::enabled`] method must be called *every* time a span
+    /// or event is recorded at that callsite.
+    ///
+    /// However, if the value returned by the [`Filter::enabled`] function will
+    /// never change for a particular [`Metadata`], regardless of the current
+    /// runtime context it's called in, then [`Filter::callsite_enabled`] may
+    /// instead return [`Interest::always`] if the callsite is enabled, or
+    /// [`Interest::never`] if it is not. This will potentially allow the
+    /// filtering result to be cached, which can improve performance, as the
+    /// filter may not need to be evaluated multiple times. Calling this method
+    /// changes the `FilterFn`'s default behavior to assume that the
+    /// [`Filter::enabled`] function *can* be cached in this manner.
```

Comment:
```suggestion
    /// [`Interest::never`] if it is not. This allows the filtering result to be 
    /// cached, which can improve performance, as the filter may not
    /// need to be evaluated multiple times. Calling this method
    /// changes the `FilterFn`'s default behavior to assume that the
    /// [`Filter::enabled`] function *can* be cached in this manner.
```

Reply:
I think the "potentially" is actually worth having, as it's not *guaranteed* to be cached; it will only be cached if other per-layer filters have the same interest, and no un-filtered layers disagree.